File: leetcode/33_Search_in_Rotated_Sorted_Array.py
import unittest
from typing import List

class Solution:
    def _binary_search(self, nums: List[int], target: int, left: int , right: int) -> int:
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] > target:
                right = mid - 1
            elif nums[mid] < target:
                left = mid + 1
            else:
                return mid
        return -1 

    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        left, right = 0, n - 1
        while left + 1 < right:
            mid = (left + right) // 2 
            if nums[mid] > nums[right]:
                left = mid
            else: 
                right = mid
        smallest = left if nums[left] < nums[right] else right

        if smallest == 0:
            left, right = 0, n - 1 
        elif target >= nums[0]:
            left, right = 0, smallest - 1
        else:
            left, right = smallest, n - 1
        
        return self._binary_search(nums, target, left, right)

class TestSearch(unittest.TestCase):
    def test_1(self):
        self.assertEqual(Solution().search([4,5,6,7,8,9,10,11,12,13,14,0,1,2], 0), 11)    

    def test_2(self):
        self.assertEqual(Solution().search([4,5,6,7,0,1,2], 3), -1)    

    def test_3(self):
        self.assertEqual(Solution().search([1], 0), -1)    

    def test_4(self):
        self.assertEqual(Solution().search([1,3], 3), 1)    

File: leetcode/33_Search_in_Rotated_Sorted_Array.rs
struct Solution {}

impl Solution {
    fn _binary_search(nums: Vec<i32>, target: i32, mut left: usize, mut right: usize) -> Option<usize> {
        while left < right {
            let mid = (left + right) / 2;
            if nums[mid] < target {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        if nums[left] == target { Some(left) } else { None }
    }

    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let n = nums.len();
        let (mut left, mut right) = (0, n - 1);
        let mut mid: usize;
        while left < right {
            mid = (left + right) / 2;
            if nums[mid] > nums[right] {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        let smallest = left;
        if smallest == 0 {
            left = 0;
            right = n - 1;
        } else if target >= nums[0] {
            left = 0;
            right = smallest - 1;
        } else {
            left = smallest;
            right = n - 1;
        }
        match Self::_binary_search(nums, target, left, right) {
            Some(index) => index as i32,
            None => -1,
        }
    }
}

#[cfg(test)]
mod search_test {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::search(vec![4,5,6,7,0,1,2], 0), 4);
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::search(vec![4,5,6,7,0,1,2], 3), -1);
    }

    #[test]
    fn test_3() {
        assert_eq!(Solution::search(vec![1], 0), -1);
    }    

    #[test]
    fn test_4() {
        assert_eq!(Solution::search(vec![3,1], 3), 0);
    }    
}

#[cfg(test)]
mod binary_search_test {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::_binary_search(vec![1], 0, 0, 0), None);
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::_binary_search(vec![1,2,3], 1, 0, 2), Some(0));
    }

    #[test]
    fn test_3() {
        assert_eq!(Solution::_binary_search(vec![1,3], 1, 0, 1), Some(0));
    }
}
